runtime/vm/createramclass.cpp

```diff
@@ -1633,53 +1643,87 @@ loadSuperClassAndInterfaces(J9VMThread *vmThread, J9ClassLoader *classLoader, J9
+			U_8 *signatureChars = J9UTF8_DATA(signature);
+			switch (signatureChars[0]) {
+			case 'Q':
+				{
+					J9Class *valueClass = internalFindClassUTF8(currentThread, signatureChars + 1, J9UTF8_LENGTH(signature) - 2, classLoader, classPreloadFlags);
+					if (NULL == valueClass) {
+						result = FALSE;
+						goto done;
+					} else {
+						J9ROMClass *valueROMClass = valueClass->romClass;
 
-				if (J9_ARE_NO_BITS_SET(valueROMClass->modifiers, J9AccValueType)) {
-					J9UTF8 *badClass = NNSRP_GET(valueROMClass->className, J9UTF8*);
-					setCurrentExceptionNLSWithArgs(currentThread, J9NLS_VM_ERROR_QTYPE_NOT_VALUE_TYPE, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR, J9UTF8_LENGTH(badClass), J9UTF8_DATA(badClass));
-					result = FALSE;
-					goto done;
-				}
+						if (J9_ARE_NO_BITS_SET(valueROMClass->modifiers, J9AccValueType)) {
+							J9UTF8 *badClass = NNSRP_GET(valueROMClass->className, J9UTF8*);
+							setCurrentExceptionNLSWithArgs(currentThread, J9NLS_VM_ERROR_QTYPE_NOT_VALUE_TYPE, J9VMCONSTANTPOOL_JAVALANGINCOMPATIBLECLASSCHANGEERROR, J9UTF8_LENGTH(badClass), J9UTF8_DATA(badClass));
+							result = FALSE;
+							goto done;
+						}
 
-				bool classIsPublic = J9_ARE_ALL_BITS_SET(valueROMClass->modifiers, J9AccPublic);
+						bool classIsPublic = J9_ARE_ALL_BITS_SET(valueROMClass->modifiers, J9AccPublic);
 
-				if ((!classIsPublic && (packageID != valueClass->packageID))
-					|| (classIsPublic && (J9_VISIBILITY_ALLOWED != checkModuleAccess(currentThread, currentThread->javaVM, romClass, module, valueROMClass, valueClass->module, valueClass->packageID, 0)))
-				) {
-					Trc_VM_CreateRAMClassFromROMClass_nestedValueClassNotVisible(currentThread, valueClass, valueClass->classLoader, classLoader);
-					setCurrentExceptionForBadClass(currentThread, J9ROMCLASS_CLASSNAME(valueROMClass), J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSERROR);
-					result = FALSE;
-					goto done;
+						if ((!classIsPublic && (packageID != valueClass->packageID))
+							|| (classIsPublic && (J9_VISIBILITY_ALLOWED != checkModuleAccess(currentThread, currentThread->javaVM, romClass, module, valueROMClass, valueClass->module, valueClass->packageID, 0)))
+						) {
+							Trc_VM_CreateRAMClassFromROMClass_nestedValueClassNotVisible(currentThread, valueClass, valueClass->classLoader, classLoader);
+							setCurrentExceptionForBadClass(currentThread, J9ROMCLASS_CLASSNAME(valueROMClass), J9VMCONSTANTPOOL_JAVALANGILLEGALACCESSERROR);
+							result = FALSE;
+							goto done;
+						}
+
+						J9FlattenedClassCache *entry = flattenedClassCache + (flattenableInstanceFieldCount + 1);
+						entry->clazz = valueClass;
+						entry->nameAndSignature = &(field->nameAndSignature);
+						entry->offset = 0;
+						flattenableInstanceFieldCount += 1;
+					}
+					/* Dont check if previously set, or in the values to avoid
+					 * extra branching, we'll only write one of them into the classFlags.
+					 */
```

Comment:
I don't understand the comment at all. Have you started using `classFlags` (a bad idea)?

Reply:
I see they are already in use. Either way, you should not be using the whole field, you need to mask the bits you actually care about (if you're masking in the caller, I suppose this isn't necessary, though it's a bit odd returning a pile of meaningless bits).